### PR TITLE
[BEAM-10529] nullable xlang coder

### PR DESCRIPTION
**Pull request by johnjcasey**
Support KafkaIO with null keys and values. The majority of this change is to enable nullable coders for xlang, and then to make minor changes to the KafkaIO itself to use them

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: model/pipeline/src/main/proto/beam_runner_api.proto
Comment (TheNeuralBit) [inline model/pipeline/src/main/proto/beam_runner_api.proto:1090]: Could you have this clarify that 1 implies non-null? It would also be nice to update the row coder doc (under "Nullable types in container types (ArrayType, MapType)") to refer to this

File: sdks/python/apache_beam/coders/typecoders_test.py
Comment (TheNeuralBit) [inline sdks/python/apache_beam/coders/typecoders_test.py:146]: nit: could you construct this with `Optional` instead of using `Union` directly?

Also I think `bytes` is already a type:
```py
>>> bytes
<class 'bytes'>
>>> type(bytes)
<class 'type'>
```

Shouldn't this be `Optional[bytes]`?

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java
Comment (TheNeuralBit) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Is this going to break the Go Kafka connector?

CC: @youngoli

**Review by TheNeuralBit — COMMENTED** [review]
Thanks @johnjcasey! I have a few suggestions

File: sdks/python/apache_beam/coders/typecoders_test.py
Comment (johnjcasey) [reply sdks/python/apache_beam/coders/typecoders_test.py:146]: Yep, I'll make the change in the test case

File: model/pipeline/src/main/proto/beam_runner_api.proto
Comment (lostluck) [reply model/pipeline/src/main/proto/beam_runner_api.proto:1090]: +1

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java
Comment (TheNeuralBit) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: :( sorry, Go didn't have a kafkaio when BEAM-10529 was filed.

Comment (chamikaramj) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Probably we can introduce a new transform that is only exposed through the Python API ?

Comment (chamikaramj) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Another option is to update URN_WITH_METADATA transforms to support nulls: https://github.com/apache/beam/blob/90dd44e81b14bd5a35f808c8962cf22cd69497c8/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L803

Seems like Go SDK uses "URN_WITHOUT_METADATA" transform: https://github.com/apache/beam/blob/90dd44e81b14bd5a35f808c8962cf22cd69497c8/sdks/go/pkg/beam/io/xlang/kafkaio/kafka.go#L84

Comment (johnjcasey) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: I think exposing the Go Nullable coder is a good option. That way we have it in the future, and we don't need to introduce work arounds in our code

Comment (TheNeuralBit) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Agreed that would be the best solution, but to be clear, are you planning on doing that for the next step on this PR?

Comment (johnjcasey) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Yep. I'm working on that now. I'm unfamiliar with Go, but making good progress so far

Comment (TheNeuralBit) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:784]: Awesome! Thank you for taking this on

File: sdks/python/apache_beam/coders/typecoders.py
Comment (chamikaramj) [inline sdks/python/apache_beam/coders/typecoders.py:143]: Did you mean to use the inner type in this call ?

File: sdks/python/apache_beam/typehints/typehints.py
Comment (chamikaramj) [inline sdks/python/apache_beam/typehints/typehints.py:510]: Does the order matter (for nullable and other cases) ?

**Review by chamikaramj — COMMENTED** [review]
Thanks!

File: sdks/python/apache_beam/coders/typecoders.py
Comment (johnjcasey) [reply sdks/python/apache_beam/coders/typecoders.py:143]: No, I thought it made more sense to use the full typehint, to try and match the existing coder.from_type_hint pattern

File: model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml
Comment (johnjcasey) [reply model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml:583]: yep. 0000 means null, 0001 means "the rest of this is non-null"

**Review by jrmccluskey — APPROVED** [review]
LGTM, just two nits

EDIT: see follow-up comment on test failures

**Comment by jrmccluskey** [comment]
R: @youngoli

File: sdks/go/pkg/beam/core/graph/coder/coder.go
Comment (damccorm) [inline sdks/go/pkg/beam/core/graph/coder/coder.go:405]: I think you may need to update https://github.com/apache/beam/blob/7e896ba429fdad0b54448409cff95bdfd1d03e92/sdks/go/pkg/beam/core/runtime/exec/coder.go#L67 (and `MakeElementDecoder`) as well to correctly plumb this through

**Comment by damccorm** [comment]
It also looks like Go CI checks are failing

**Comment by jrmccluskey** [comment]
RAT precommit is tied to files not having the Apache license header, TestStandardCoders failure in the Go Precommit looks like it's tied to the coder needing to be plumbed through exec/coder.go as Danny pointed out

**Comment by damccorm** [comment]
Should we be adding this here as well? https://github.com/apache/beam/blob/2a23166f738182f4553a79b8fc44c46e10705e76/sdks/go/test/regression/coders/fromyaml/fromyaml.go#L210

**Comment by johnjcasey** [comment]
run java precommit

**Comment by johnjcasey** [comment]
run java precommit

**Comment by johnjcasey** [comment]
run java precommit

**Review by damccorm — APPROVED** [review]
I left some comments about the tests done, but it mostly looks good to me. I would like someone with more experience with go coders to take a look as well though.

R: @lostluck - would you mind taking a look?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslators.java
Comment (chamikaramj) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslators.java:211]: Raise an error if there's more than one component coder ?

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/wire/CommonCoderTest.java
Comment (chamikaramj) [inline runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/wire/CommonCoderTest.java:387]: Are we generating individual tests for coders here or are we lumping everything into a single test ? If it's latter probably we should consider adding tests specifically for the new case.

File: sdks/go/pkg/beam/core/graph/coder/coder.go
Comment (chamikaramj) [reply sdks/go/pkg/beam/core/graph/coder/coder.go:407]: Pls address this.

File: sdks/python/apache_beam/coders/coders.py
Comment (chamikaramj) [inline sdks/python/apache_beam/coders/coders.py:581]: You just expect one inner type here as well ?
If so we should probably validate that and simplify this code.

File: sdks/python/apache_beam/coders/coders.py
Comment (chamikaramj) [inline sdks/python/apache_beam/coders/coders.py:597]: Do we have tests to make sure that coders of the three languages are compatible ? For example, that you can encode a value of type "NullableCoder(Other StandardCoder)" from SDK1 and  decode from SDK2 and vice versa for all SDK combinations ?

@TheNeuralBit might know the answer.

File: sdks/python/apache_beam/coders/standard_coders_test.py
Comment (chamikaramj) [inline sdks/python/apache_beam/coders/standard_coders_test.py:196]: Hmm, why did we have to update this existing test ? (just to make sure that there's no backwards compatibility issue here).

File: sdks/python/apache_beam/typehints/typehints.py
Comment (chamikaramj) [inline sdks/python/apache_beam/typehints/typehints.py:607]: May be call this "is_nullable" to  be more specific ?

File: sdks/python/apache_beam/typehints/typehints_test.py
Comment (chamikaramj) [inline sdks/python/apache_beam/typehints/typehints_test.py:323]: Add more tests to check encoding/decoding with NullableCoder ? (we may already have such tests)

**Review by chamikaramj — CHANGES_REQUESTED** [review]
Thanks!

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/wire/CommonCoderTest.java
Comment (johnjcasey) [reply runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/wire/CommonCoderTest.java:387]: It is a single test in the java sense, but there are 3 test cases in the yaml file for nullable coders.

File: sdks/go/pkg/beam/core/graph/coder/coder.go
Comment (johnjcasey) [reply sdks/go/pkg/beam/core/graph/coder/coder.go:407]: Added

File: sdks/python/apache_beam/coders/coders.py
Comment (johnjcasey) [reply sdks/python/apache_beam/coders/coders.py:597]: Yes. We have a yaml file that contains test cases that all three sdks read and validate. org/apache/beam/model/fnexecution/v1/standard_coders.yaml

File: sdks/python/apache_beam/coders/standard_coders_test.py
Comment (johnjcasey) [reply sdks/python/apache_beam/coders/standard_coders_test.py:196]: This is the python file that actually reads the yaml, and runs the tests. I needed to register the nullable coder in it, so it knows what the nullable urn points to

File: sdks/python/apache_beam/typehints/typehints.py
Comment (johnjcasey) [reply sdks/python/apache_beam/typehints/typehints.py:607]: good idea

File: sdks/python/apache_beam/typehints/typehints_test.py
Comment (johnjcasey) [reply sdks/python/apache_beam/typehints/typehints_test.py:323]: We are, Brian wrote them when NullableCoder was just used for nested encoding

**Review by chamikaramj — APPROVED** [review]
LGTM. Thanks.

**Comment by johnjcasey** [comment]
Waiting on @lostluck to review the Go changes, but then I am good to go

File: sdks/go/pkg/beam/core/runtime/exec/coder.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/coder.go:665]: Please have these methods on pointer receivers, not the value receivers. (*nullableDecoder here and *nullableEncoder above). This keeps them consistent with the rest of the file, which helps readability, and avoids  unnecessary usage bugs.

File: sdks/go/test/regression/coders/fromyaml/fromyaml.go
Comment (lostluck) [reply sdks/go/test/regression/coders/fromyaml/fromyaml.go:225]: This code has design problems, but they pre-exist John's addition, so he's largely following the existing structure. In this case, this was an unscheduled task that I had to do in a hurry, with little oversight, and improving it has never been a priority. This can certainly be improved significantly, but likely in a separate PR.

So generally, I agree with Danny here. But specifically...

1. As this is specific purpose comparison code, it is true it may blow up, but it will only blow up with a runtime type assertion error when someone is adding new/strange cases to it. The [defer'd recover](https://github.com/apache/beam/blob/488e1d165e121d757fac8b9260d004de93dbd338/sdks/go/test/regression/coders/fromyaml/fromyaml.go#L122) at the top of the encoder case loop will narrow down where the problem is, if not the actual example that blew up, which should be enough information to determine the offending types. 

2. Agreed, in this case, there's probably a separate bug going on. But this is how the plain `bytes` coder is handling it. IIRC there was something very odd with the YAML parsing which lead to flaky interpretations of the types. It was easier to work around it at the time.

3. In this case, I disagree: Downstream the `want` and `got` are compared with `cmp`, and that will produce a much clearer error message than what we could add here, along with the full coder information.

**Review by lostluck — CHANGES_REQUESTED** [review]
A few quick fixes to the Go changes, but otherwise, I like it. Thank you!

**Review by lostluck — APPROVED** [review]
LGTM for Go. Thanks!

**Comment by lostluck** [comment]
The GoPortable error is xlang related but not due to this PR.

**Comment by johnjcasey** [comment]
run java precommit

**Comment by johnjcasey** [comment]
Run Portable_Python PreCommit

**Comment by johnjcasey** [comment]
Failures are due to https://issues.apache.org/jira/browse/BEAM-14231, and this can be merged

**Comment by TheNeuralBit** [comment]
Re-running the Python failure, looks like it should be resolved by rollback #17255. I will merge when we get it green

@lostluck is there a jira for the Go xlang failure?

**Comment by TheNeuralBit** [comment]
> Re-running the Python failure, looks like it should be resolved by rollback #17255. I will merge when we get it green

Oh I forgot, GHA checks do not run with changes from HEAD, so it's still running code with this flaky behavior.

**Comment by jrmccluskey** [comment]
@johnjcasey The Actions build breakage was due to bad formatting in core/typex/special.go. I've got a PR out to fix it (linked above) but please don't ignore the build status moving forward!

**Comment by TheNeuralBit** [comment]
@jrmccluskey This was my fault, I assumed the Go failure was the one @lostluck was referencing [here](https://github.com/apache/beam/pull/16923#issuecomment-1084793890).

**Comment by Abacn** [comment]
Hi,

I got an error on `./gradlew :sdks:python:wordCount` checking out latest master:

```> Task :sdks:python:wordCount FAILED
Traceback (most recent call last):
  File "/Users/<USER>/.pyenv/versions/3.8.7/lib/python3.8/runpy.py", line 185, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/Users/<USER>/.pyenv/versions/3.8.7/lib/python3.8/runpy.py", line 111, in _get_module_details
    __import__(pkg_name)
  File "/Users/<USER>/beam/sdks/python/apache_beam/__init__.py", line 92, in <module>
    from apache_beam import coders
  File "/Users/<USER>/beam/sdks/python/apache_beam/coders/__init__.py", line 17, in <module>
    from apache_beam.coders.coders import *
  File "/Users/<USER>/beam/sdks/python/apache_beam/coders/coders.py", line 638, in <module>
    Coder.register_structured_urn(common_urns.coders.NULLABLE.urn, NullableCoder)
AttributeError: type object 'Enum' has no attribute 'NULLABLE'

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/<USER>/beam/sdks/python/build.gradle' line: 119

* What went wrong:
Execution failed for task ':sdks:python:wordCount'.
> Process 'command 'sh'' finished with non-zero exit value 1
```

Seems related to this change. Could anyone please have a look, thanks!

**Comment by Abacn** [comment]
> Hi,
> 
> I got an error on `./gradlew :sdks:python:wordCount` checking out latest master:
> 
> ```
> Traceback (most recent call last):
>   File "/Users/<USER>/.pyenv/versions/3.8.7/lib/python3.8/runpy.py", line 185, in _run_module_as_main
>     mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
>   File "/Users/<USER>/.pyenv/versions/3.8.7/lib/python3.8/runpy.py", line 111, in _get_module_details
>     __import__(pkg_name)
>   File "/Users/<USER>/beam/sdks/python/apache_beam/__init__.py", line 92, in <module>
>     from apache_beam import coders
>   File "/Users/<USER>/beam/sdks/python/apache_beam/coders/__init__.py", line 17, in <module>
>     from apache_beam.coders.coders import *
>   File "/Users/<USER>/beam/sdks/python/apache_beam/coders/coders.py", line 638, in <module>
>     Coder.register_structured_urn(common_urns.coders.NULLABLE.urn, NullableCoder)
> AttributeError: type object 'Enum' has no attribute 'NULLABLE'
> 
> FAILURE: Build failed with an exception.
> 
> * Where:
> Build file '/Users/<USER>/beam/sdks/python/build.gradle' line: 119
> 
> * What went wrong:
> Execution failed for task ':sdks:python:wordCount'.
> > Process 'command 'sh'' finished with non-zero exit value 1
> ```
> 
> Seems related to this change. Could anyone please have a look, thanks!

It turns out that this was a problem of my protobuf dependency, which causes sdks/python/gen_protos.py not running to generate new versions of RegularCoders.py. Force install protobuf with c++ implementation resolves this problem.

**Pull request merged**
